### PR TITLE
finish trait solver skeleton work

### PR DESCRIPTION
**Pull request by lcnr**
### 648d661b4e0fcf55f7082894f577377eb451db4b

The previous implementation didn't remove provisional entries which depended on the current goal if we're forced to rerun in case the provisional result of that entry is different from the new result. For reference, see https://rust-lang.github.io/chalk/book/recursive/search_graph.html. 

We should also treat inductive cycles as overflow, not ordinary ambiguity.

### 219a5de2517cebfe20a2c3417bd302f7c12db70c 6a1912be539dd5a3b3c10be669787c4bf0c1868a

These two commits move canonicalization to the start of the queries which simplifies a bunch of stuff. I originally intended to keep stuff canonicalized for a while because I expected us to add a additional caches the trait solver, either for candidate assembly or for projections. We ended up not adding (and expect to not need) any of them so this just ends up being easier to understand.

### d78d5ad0979e965afde6500bccfa119b47063506

adds a special `eq` for the solver which doesn't care about obligations or spans

### 18704e6a78b7703e1bbb3856f015cb76c0a07a06

implements https://rust-lang.zulipchat.com/#narrow/stream/364551-t-types.2Ftrait-system-refactor/topic/projection.20cache

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_trait_selection/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ rustc_infer = { path = "../rustc_infer" }
 rustc_lint_defs = { path = "../rustc_lint_defs" }
 rustc_macros = { path = "../rustc_macros" }
 rustc_query_system = { path = "../rustc_query_system" }
+rustc_serialize = { path = "../rustc_serialize" }
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/Cargo.toml:22]: Not blocking, but but what's this crate dependency for?

Comment (lcnr) [reply compiler/rustc_trait_selection/Cargo.toml:22]: for `rustc_index::newtype_index!`

**Comment by albertlarsan68** [comment]
@rusbot label -A-bootstrap

File: compiler/rustc_trait_selection/src/solve/project_goals.rs [hunk 2]
```diff
@@ -11,19 +11,112 @@ use rustc_infer::traits::query::NoSolution;
...
+use std::ops::ControlFlow;
 
 impl<'tcx> EvalCtxt<'_, 'tcx> {
     pub(super) fn compute_projection_goal(
         &mut self,
         goal: Goal<'tcx, ProjectionPredicate<'tcx>>,
     ) -> QueryResult<'tcx> {
-        let candidates = self.assemble_and_evaluate_candidates(goal);
-        self.merge_project_candidates(candidates)
+        // To only compute normalization ones for each projection we only
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/project_goals.rs:26]: "to only compute *ones*", fixme: `once`.

**Pull request merged**